src/OpenTelemetry.Exporter.Jaeger/JaegerExporter.cs

```diff
@@ -113,10 +147,53 @@ internal void ApplyLibraryResource(Resource libraryResource)
             }
         }
 
-        /// <inheritdoc/>
-        protected override void OnShutdown(int timeoutMilliseconds)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void AppendSpan(JaegerSpan jaegerSpan)
         {
-            _ = this.JaegerAgentUdpBatcher.FlushAsync(default).GetAwaiter().GetResult();
+            if (this.processCache == null)
+            {
+                this.Process.Message = this.BuildThriftMessage(this.Process).ToArray();
+                this.processCache = new Dictionary<string, Process>
+                {
+                    [this.Process.ServiceName] = this.Process,
+                };
+            }
+
+            var spanServiceName = jaegerSpan.PeerServiceName ?? this.Process.ServiceName;
+
+            if (!this.processCache.TryGetValue(spanServiceName, out var spanProcess))
+            {
+                spanProcess = new Process(spanServiceName, this.Process.Tags);
+                spanProcess.Message = this.BuildThriftMessage(spanProcess).ToArray();
+                this.processCache.Add(spanServiceName, spanProcess);
```

Comment:
Curious - is there a need to put an upper bound (something like splay tree with a max size) or we are pretty confident that there won't be too many unique service names?

Reply:
Not sure. Kind of related to #1235. I feel like it would be a special/rare case where you would be connecting to an ever-growing list of services. Some kind of gateway thing perhaps?